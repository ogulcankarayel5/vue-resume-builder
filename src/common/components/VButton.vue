<template>
  <button
    data-test-id="button"
    class="py-4 flex justify-center items-center"
    :class="[getVariant, disabled, getSize, getElevation, getRounded]"
    @click="onClick"
  >
    <pulse-loader data-test-id="loading" v-if="loading" color="#ffffff" />
    <slot v-else data-test-id="slot" />
  </button>
</template>

<script lang="ts">
import { defineComponent } from "vue";
type Variants = "primary";
type Size = "block" | "medium";

export default defineComponent({
  name: "VButton",
  props: {
    variant: {
      type: String as () => Variants,
      required: false,
    },
    size: {
      type: String as () => Size,
      default: "block",
    },
    loading: {
      type: Boolean,
      default: false,
    },
    elevation: {
      type: Boolean,
      default: false,
    },
    rounded: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    getSize(): any {
      switch (this.size) {
        case "block":
          return "w-full";
        case "medium":
          return "w-36";
        default:
          break;
      }
      return "w-100";
    },

    getVariant(): string {
      switch (this.variant) {
        case "primary":
          return "bg-primary text-white font-semibold ";
        default:
          break;
      }

      return "";
    },
    disabled() {
      return {
        "bg-opacity-50 cursor-default": this.$attrs.disabled,
      };
    },
    getElevation() {
      return {
        "shadow-2xl": this.elevation,
      };
    },
    getRounded() {
      return {
        "rounded-lg": this.rounded,
      };
    },
  },
  methods: {
    onClick(event: Event) {
      this.$emit("onClick", event);
    },
  },
});
</script>

<style></style>
