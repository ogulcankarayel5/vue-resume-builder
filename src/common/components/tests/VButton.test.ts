import VButton from "@/common/components/VButton.vue";
import { shallowMount } from "@vue/test-utils";

const factory = (values = {}, other = {}) => {
  return shallowMount(VButton, {
    data() {
      return {
        ...values,
      };
    },
    ...other,
  });
};

describe("VButton", () => {
  it("snapshot", () => {
    const wrapper = factory();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("button click", () => {
    const wrapper = factory();
    wrapper.find('[data-test-id="button"]').trigger("click");
    expect(wrapper.emitted().onClick).toBeTruthy();
  });
  it("display loading if loading is true", () => {
    const wrapper = factory(
      {},
      {
        propsData: {
          loading: true,
        },
        slots: {
          default: "<span>test</span>",
        },
      }
    );
    expect(wrapper.find('[data-test-id="loading"]').exists()).toBe(true);
    expect(wrapper.find("span").exists()).toBe(false);
  });
  it("display slot if loading is false", () => {
    const wrapper = factory(
      {},
      {
        propsData: {
          loading: false,
        },
        slots: {
          default: "<span>test</span>",
        },
      }
    );
    expect(wrapper.find('[data-test-id="loading"]').exists()).toBe(false);
    expect(wrapper.find("span").text()).toBe("test");
  });

  describe("variant", () => {
    const variants = ["primary"];
    variants.forEach((item) => {
      it(`should render ${item} variant class`, () => {
        const wrapper = factory(
          {},
          {
            propsData: {
              variant: item,
            },
          }
        );
        expect(wrapper.find('[data-test-id="button"]').classes()).toContain(
          `bg-${item}`
        );
      });
    });
  });
});
