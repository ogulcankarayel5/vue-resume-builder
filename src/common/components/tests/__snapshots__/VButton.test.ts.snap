// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VButton snapshot 1`] = `
<button
  class="py-4 flex justify-center items-center w-full"
  data-test-id="button"
/>
`;
